### ISSUE TITLE
Serverside chat.AddText

### ISSUE THREAD

**Issue by Leystryku** [Facepunch/garrysmod-requests#122]
It would be really neat to have a serverside version of chat.AddText. Parameters should be the same ones as the client's chat.AddText has, but with the possibility to use a table or player as first argument ( If no table or player is given as first argument, send to all players ). The Serverside version would be really useful, because right now there is no function included in GMod to send colored text messages to clients.

**Comment by edgarasf123** [comment]
We really need this, I have seen multiple addons creating multiple function which does the same thing. I even seen stuff like this
```ply:SendLua("chat.AddText(Color(40,40,200),\"[RTD]\",Color(200,200,20),\" "..s1.."\")"); ```
Or for example if I have serverside only addon, and I want to print something to players, I have to use because I don't see the point of creating seperate file just for one function.
```
hook.Add( "PlayerInitialSpawn", "DonationSystemPlayerJoinSendLua", function(ply)
	ply:SendLua( [[ 
		net.Receive( "DonationSystemColorChat", function( len ) chat.AddText( unpack( net.ReadTable() ) ) end ) 
	]] )
end)
```

**Comment by Leystryku** [comment]
>2018
>still waiting
wtf?

**Comment by MrUnknownGamer** [comment]
bump

**Comment by nemesis64** [comment]
>2019
>still waiting

**Comment by Bluscream** [comment]
> 2020
> still waiting

**Comment by Deyvan** [comment]
>2021
still waiting

**Comment by GitSparTV** [comment]
> Just make your own

**Comment by robotboy655** [comment]
See https://github.com/Facepunch/garrysmod/pull/1333 for reasoning.

**Comment by thegrb93** [comment]
So basically still "Implement your own"

**Comment by Kefta** [comment]
No, "implement your own" implies you should be doing it - you shouldn't. You should cache the constant strings and colours on the client.

**Comment by thegrb93** [comment]
Not possible for a lot of cases.

**Comment by thegrb93** [comment]
E2 for example.

**Comment by h3xcat** [comment]
It would make sense to cache constant strings if messages were printed a thousand times a second. But we're talking about chat messages, messages that are printed a few times a minute at most. It's a bit overkill to optimize such functionality.

**Comment by thegrb93** [comment]
^ messages that are quite dynamic too

**Comment by Kefta** [comment]
E2 should add their own solution to their library then. And if the string is dynamic, cache only the constant parts on the client then send the dynamic portions over the network. Better yet, those dynamic portions can usually be mapped to a string table so instead of sending a string over the network, you send a uint.

**Comment by h3xcat** [comment]
So you'd rather depend on people to write optimized code? I can guarantee that at least 90% of developers are not going to utilize your strategy and end up writing something that is less efficient than what you'd create for dynamic strings.

**Comment by h3xcat** [comment]
Even ulib doesn't use constant strings, which is probably the most utilized library in gmod. And it makes sense cause it would complicate the code quite a bit.

https://github.com/TeamUlysses/ulib/blob/e887b4086124a69fad9efd58ce6ce4311d60e3aa/lua/ulib/shared/messages.lua#L159

**Comment by GitSparTV** [comment]
> Even ulib doesn't use constant strings

Well https://github.com/Facepunch/garrysmod/pull/1333 doesn't have string caching as well

**Comment by thegrb93** [comment]
For fuck's sake. https://github.com/wiremod/wire/pull/2157/files#diff-b21127288048d7221b5ae884a93b665d1149b0c0289ca3221671bd85b2e6f89aR29-R45

**Comment by Derpius** [comment]
**TL/DR:** The arguments against implementing a `net` library based `ChatPrint` replacement or serverside `chat.AddText` are weak at best, having to remake the same function over and over again is ridiculous.  

---

Came here to make a PR to add a not completely outdated umsg based chat print cause I found myself having to implement it yet again for some server utils, as expected other people had tried in the past to no avail, and then I find my Wire PR's `BetterChatPrint` code linked.  

Why is this still not a thing, and as for "not sending dynamic strings over net", the use case I have for my utils, and many other uses cases **require** either partially, or entirely dynamic strings to be used. (for instance the E2 chatprint example shown above).  

And if you're worried about implementing a do-all completely dynamic function causing people who *would've* implemented an optimised solution using that instead, this is gmod, no one is gonna do that, hell you'll likely end up with less optimised solutions being used than the best dynamic solution.  
For instance my `BetterChatPrint` in wire was done in the shortest time possible cause I'm tired of having to reinvent the wheel over and over again, so it's not even compressing the strings before sending, never mind saving static message formats on the client.  

Lastly, I highly doubt the current `ChatPrint` implements any kind of caching, so replacing it with `net` and just setting say, a 1024 to 4096 char limit (anything higher should be using a custom implementation, probably with streaming) would render any points about optimisation almost completely mute, a couple thousand chars isn't gonna make a dent.
